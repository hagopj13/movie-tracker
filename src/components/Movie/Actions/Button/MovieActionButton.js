// @flow
import React from 'react';
import type { Node } from 'react';
import { makeStyles } from '@material-ui/core/styles';
import Tooltip from '@material-ui/core/Tooltip';
import Fab from '@material-ui/core/Fab';
import { lightBlue } from '@material-ui/core/colors';

type Props = {
  children: Node,
  title: string,
  onClick: () => void,
};

const useStyles = makeStyles((theme) => ({
  tooltip: {
    fontSize: theme.typography.subtitle2.fontSize,
  },
  fab: {
    backgroundColor: lightBlue[800],
    '&:hover': {
      backgroundColor: lightBlue[900],
    },
  },
}));

const MovieActionButton = (props: Props, ref) => {
  const { children, title, onClick } = props;

  const classes = useStyles();

  return (
    <Tooltip classes={{ tooltip: classes.tooltip }} title={title}>
      <Fab className={classes.fab} ref={ref} size="medium" onClick={onClick}>
        {children}
      </Fab>
    </Tooltip>
  );
};

export default React.forwardRef<Props, Fab>(MovieActionButton);
