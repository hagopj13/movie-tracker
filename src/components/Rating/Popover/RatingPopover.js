// @flow
import React, { useState } from 'react';
import { makeStyles } from '@material-ui/core/styles';
import Popover from '@material-ui/core/Popover';
import Box from '@material-ui/core/Box';
import Rating from '@material-ui/lab/Rating';

type Props = {
  isOpen: boolean,
  anchorEl: HTMLElement | null,
  initialValue: number,
  onClose: () => void,
  onChange: (newValue: number | null) => void,
};

const useStyles = makeStyles((theme) => ({
  content: {
    display: 'flex',
    alignItems: 'center',
    padding: theme.spacing(1.5),
    width: 165,
  },
}));

const RatingPopover = (props: Props) => {
  const { isOpen, anchorEl, initialValue, onClose, onChange } = props;

  const [value, setValue] = useState(initialValue);
  const [hover, setHover] = useState(-1);

  const classes = useStyles();

  const handleChange = (event, newValue: number | null) => {
    setValue(newValue);
    onChange(newValue);
  };

  const handleChangeActive = (event, newHover: number) => {
    setHover(newHover);
  };

  return (
    <Popover
      open={isOpen}
      anchorEl={anchorEl}
      onClose={onClose}
      anchorOrigin={{ vertical: 'bottom', horizontal: 'center' }}
      transformOrigin={{ vertical: -10, horizontal: 'center' }}
    >
      <div className={classes.content}>
        <Rating
          name="rating"
          value={value}
          precision={0.5}
          onChange={handleChange}
          onChangeActive={handleChangeActive}
        />
        <Box component="span" ml={1} fontSize={15}>
          {hover !== -1 ? hover : value}
        </Box>
      </div>
    </Popover>
  );
};

export default RatingPopover;
