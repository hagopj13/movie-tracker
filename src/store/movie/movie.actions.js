// @flow
import type { Movie, MovieUserState } from 'types';

import MovieActionTypes from './movie.types';

const fetchMovie = (id: string) => ({
  type: MovieActionTypes.FETCH_MOVIE,
  payload: { id },
});

const fetchMovieStart = () => ({
  type: MovieActionTypes.FETCH_MOVIE_START,
});

const fetchMovieSuccess = (movie: Movie, movieUserState: MovieUserState) => ({
  type: MovieActionTypes.FETCH_MOVIE_SUCCESS,
  payload: { movie, movieUserState },
});

const fetchMovieFailure = (error: string) => ({
  type: MovieActionTypes.FETCH_MOVIE_FAILURE,
  payload: { error },
});

const fetchMovieUserState = (id: string) => ({
  type: MovieActionTypes.FETCH_MOVIE_USER_STATE,
  payload: { id },
});

const fetchMovieUserStateStart = () => ({
  type: MovieActionTypes.FETCH_MOVIE_USER_STATE_START,
});

const fetchMovieUserStateSuccess = (movieUserState: MovieUserState) => ({
  type: MovieActionTypes.FETCH_MOVIE_USER_STATE_SUCCESS,
  payload: { movieUserState },
});

const fetchMovieUserStateFailure = (error: string) => ({
  type: MovieActionTypes.FETCH_MOVIE_USER_STATE_FAILURE,
  payload: { error },
});

const setIsMovieFavorite = (id: number, isFavorite: boolean) => ({
  type: MovieActionTypes.SET_IS_MOVIE_FAVORITE,
  payload: { id, isFavorite },
});

const setIsMovieFavoriteStart = () => ({
  type: MovieActionTypes.SET_IS_MOVIE_FAVORITE_START,
});

const setIsMovieFavoriteSuccess = (id: number, isFavorite: boolean) => ({
  type: MovieActionTypes.SET_IS_MOVIE_FAVORITE_SUCCESS,
  payload: { id, isFavorite },
});

const setIsMovieFavoriteFailure = (error: string) => ({
  type: MovieActionTypes.SET_IS_MOVIE_FAVORITE_FAILURE,
  payload: { error },
});

const setIsMovieInWatchlist = (id: string, isInWatchlist: boolean) => ({
  type: MovieActionTypes.SET_IS_MOVIE_IN_WATCHLIST,
  payload: { id, isInWatchlist },
});

const setIsMovieInWatchlistStart = () => ({
  type: MovieActionTypes.SET_IS_MOVIE_IN_WATCHLIST_START,
});

const setIsMovieInWatchlistSuccess = (id: number, isInWatchlist: boolean) => ({
  type: MovieActionTypes.SET_IS_MOVIE_IN_WATCHLIST_SUCCESS,
  payload: { id, isInWatchlist },
});

const setIsMovieInWatchlistFailure = (error: string) => ({
  type: MovieActionTypes.SET_IS_MOVIE_IN_WATCHLIST_FAILURE,
  payload: { error },
});

const rateMovie = (id: number, rating: number | null) => ({
  type: MovieActionTypes.RATE_MOVIE,
  payload: { id, rating },
});

const rateMovieStart = () => ({
  type: MovieActionTypes.RATE_MOVIE_START,
});

const rateMovieSuccess = (id: number, rating: number | null) => ({
  type: MovieActionTypes.RATE_MOVIE_SUCCESS,
  payload: { id, rating },
});

const rateMovieFailure = (error: string) => ({
  type: MovieActionTypes.RATE_MOVIE_FAILURE,
  payload: { error },
});

export default {
  fetchMovie,
  fetchMovieStart,
  fetchMovieSuccess,
  fetchMovieFailure,
  fetchMovieUserState,
  fetchMovieUserStateStart,
  fetchMovieUserStateSuccess,
  fetchMovieUserStateFailure,
  setIsMovieFavorite,
  setIsMovieFavoriteStart,
  setIsMovieFavoriteSuccess,
  setIsMovieFavoriteFailure,
  setIsMovieInWatchlist,
  setIsMovieInWatchlistStart,
  setIsMovieInWatchlistSuccess,
  setIsMovieInWatchlistFailure,
  rateMovie,
  rateMovieStart,
  rateMovieSuccess,
  rateMovieFailure,
};
