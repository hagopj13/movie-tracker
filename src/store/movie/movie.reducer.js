// @flow
import { handleActions } from 'redux-actions';

import AuthActionTypes from 'store/auth/auth.types';
import type { Movie, MovieUserState } from 'types';

import MovieActionTypes from './movie.types';

type State = {
  details: Movie,
  userState: MovieUserState,
};

export const initialState: State = {
  details: {
    id: null,
    title: '',
    releaseDate: '',
    voteAverage: 0,
    voteCount: 0,
    posterPath: '',
    backdropPath: '',
    tagline: '',
    overview: '',
    runtime: 0,
    status: '',
    genres: [],
    actors: [],
    reviews: [],
    recommendations: [],
  },
  userState: {
    isFavorite: false,
    isInWatchlist: false,
    rating: null,
  },
};

const fetchMovieStart = () => initialState;

const fetchMovieSuccess = (state: State, action) => ({
  ...state,
  details: action.payload.movie,
  userState: action.payload.movieUserState,
});

const fetchMovieUserStateSuccess = (state: State, action) => ({
  ...state,
  userState: action.payload.movieUserState,
});

const setIsMovieFavoriteSuccess = (state: State, action) => {
  if (action.payload.id === state.details?.id) {
    return {
      ...state,
      userState: {
        ...state.userState,
        isFavorite: action.payload.isFavorite,
      },
    };
  }
  return state;
};

const setIsMovieInWatchlistSuccess = (state: State, action) => {
  if (action.payload.id === state.details?.id) {
    return {
      ...state,
      userState: {
        ...state.userState,
        isInWatchlist: action.payload.isInWatchlist,
      },
    };
  }
  return state;
};

const rateMovieSuccess = (state: State, action) => {
  if (action.payload.id === state.details?.id) {
    return {
      ...state,
      userState: {
        ...state.userState,
        rating: action.payload.rating,
      },
    };
  }
  return state;
};

const logoutSuccess = (state: State) => ({
  ...state,
  userState: initialState.userState,
});

const movieActionHandler = {
  [MovieActionTypes.FETCH_MOVIE_START]: fetchMovieStart,
  [MovieActionTypes.FETCH_MOVIE_SUCCESS]: fetchMovieSuccess,
  [MovieActionTypes.FETCH_MOVIE_USER_STATE_SUCCESS]: fetchMovieUserStateSuccess,
  [MovieActionTypes.SET_IS_MOVIE_FAVORITE_SUCCESS]: setIsMovieFavoriteSuccess,
  [MovieActionTypes.SET_IS_MOVIE_IN_WATCHLIST_SUCCESS]: setIsMovieInWatchlistSuccess,
  [MovieActionTypes.RATE_MOVIE_SUCCESS]: rateMovieSuccess,
  [AuthActionTypes.LOGOUT_SUCCESS]: logoutSuccess,
};

export default handleActions(movieActionHandler, initialState);
