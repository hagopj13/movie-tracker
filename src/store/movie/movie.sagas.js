import { all, call, takeLatest, put, select } from 'redux-saga/effects';

import * as api from 'api/tmdb';
import authSelectors from 'store/auth/auth.selectors';
import { convertResponseToMovie, convertResponseToMovieUserState } from 'api/tmdb/utils';

import MovieActionTypes from './movie.types';
import movieActions from './movie.actions';

export function* fetchMovie({ payload: { id } }) {
  yield put(movieActions.fetchMovieStart());
  const sessionId = yield select(authSelectors.selectSessionId);
  try {
    const { data } = yield call(api.getMovieDetails, id, sessionId);
    const movie = yield call(convertResponseToMovie, data);
    const movieUserState = yield call(convertResponseToMovieUserState, data.account_states);
    yield put(movieActions.fetchMovieSuccess(movie, movieUserState));
  } catch (error) {
    yield put(movieActions.fetchMovieFailure(error.status_message));
  }
}

export function* fetchMovieUserState({ payload: { id } }) {
  yield put(movieActions.fetchMovieUserStateStart());
  const sessionId = yield select(authSelectors.selectSessionId);
  try {
    const { data } = yield call(api.getMovieUserState, id, sessionId);
    const movieUserState = yield call(convertResponseToMovieUserState, data);
    yield put(movieActions.fetchMovieUserStateSuccess(movieUserState));
  } catch (error) {
    yield put(movieActions.fetchMovieUserStateFailure(error.status_message));
  }
}

export function* setIsMovieFavorite({ payload: { id, isFavorite } }) {
  yield put(movieActions.setIsMovieFavoriteStart());
  const sessionId = yield select(authSelectors.selectSessionId);
  const accountId = yield select(authSelectors.selectAccountId);
  try {
    yield call(api.setIsMovieFavorite, id, isFavorite, sessionId, accountId);
    yield put(movieActions.setIsMovieFavoriteSuccess(id, isFavorite));
  } catch (error) {
    yield put(movieActions.setIsMovieFavoriteFailure(error.status_message));
  }
}

export function* setIsMovieInWatchlist({ payload: { id, isInWatchlist } }) {
  yield put(movieActions.setIsMovieInWatchlistStart());
  const sessionId = yield select(authSelectors.selectSessionId);
  const accountId = yield select(authSelectors.selectAccountId);
  try {
    yield call(api.setIsMovieInWatchlist, id, isInWatchlist, sessionId, accountId);
    yield put(movieActions.setIsMovieInWatchlistSuccess(id, isInWatchlist));
  } catch (error) {
    yield put(movieActions.setIsMovieInWatchlistFailure(error.status_message));
  }
}

export function* rateMovie({ payload: { id, rating } }) {
  yield put(movieActions.rateMovieStart());
  const sessionId = yield select(authSelectors.selectSessionId);
  try {
    if (rating) {
      yield call(api.rateMovie, id, rating * 2, sessionId);
    } else {
      yield call(api.deleteMovieRating, id, sessionId);
    }
    yield put(movieActions.rateMovieSuccess(id, rating));
  } catch (error) {
    yield put(movieActions.rateMovieFailure(error.status_message));
  }
}

export function* onFetchMovie() {
  yield takeLatest(MovieActionTypes.FETCH_MOVIE, fetchMovie);
}

export function* onFetchMovieUserState() {
  yield takeLatest(MovieActionTypes.FETCH_MOVIE_USER_STATE, fetchMovieUserState);
}

export function* onSetIsMovieFavorite() {
  yield takeLatest(MovieActionTypes.SET_IS_MOVIE_FAVORITE, setIsMovieFavorite);
}

export function* onSetIsMovieInWatchlist() {
  yield takeLatest(MovieActionTypes.SET_IS_MOVIE_IN_WATCHLIST, setIsMovieInWatchlist);
}

export function* onRateMovie() {
  yield takeLatest(MovieActionTypes.RATE_MOVIE, rateMovie);
}

export default function* movieSagas() {
  yield all([
    call(onFetchMovie),
    call(onFetchMovieUserState),
    call(onSetIsMovieFavorite),
    call(onSetIsMovieInWatchlist),
    call(onRateMovie),
  ]);
}
