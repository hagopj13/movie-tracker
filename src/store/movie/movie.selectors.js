// @flow
import { createSelector } from 'reselect';

const selectMovie = (state) => state.movie;

const selectMovieDetails = createSelector([selectMovie], (movie) => movie.details);

const selectMovieUserState = createSelector([selectMovie], (movie) => movie.userState);

export default {
  selectMovieDetails,
  selectMovieUserState,
};
