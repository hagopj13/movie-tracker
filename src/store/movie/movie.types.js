export default {
  FETCH_MOVIE: 'FETCH_MOVIE',
  FETCH_MOVIE_START: 'FETCH_MOVIE_START',
  FETCH_MOVIE_SUCCESS: 'FETCH_MOVIE_SUCCESS',
  FETCH_MOVIE_FAILURE: 'FETCH_MOVIE_FAILURE',
  FETCH_MOVIE_USER_STATE: 'FETCH_MOVIE_USER_STATE',
  FETCH_MOVIE_USER_STATE_START: 'FETCH_MOVIE_USER_STATE_START',
  FETCH_MOVIE_USER_STATE_SUCCESS: 'FETCH_MOVIE_USER_STATE_SUCCESS',
  FETCH_MOVIE_USER_STATE_FAILURE: 'FETCH_MOVIE_USER_STATE_FAILURE',
  SET_IS_MOVIE_FAVORITE: 'SET_IS_MOVIE_FAVORITE',
  SET_IS_MOVIE_FAVORITE_START: 'SET_IS_MOVIE_FAVORITE_START',
  SET_IS_MOVIE_FAVORITE_SUCCESS: 'SET_IS_MOVIE_FAVORITE_SUCCESS',
  SET_IS_MOVIE_FAVORITE_FAILURE: 'SET_IS_MOVIE_FAVORITE_FAILURE',
  SET_IS_MOVIE_IN_WATCHLIST: 'SET_IS_MOVIE_IN_WATCHLIST',
  SET_IS_MOVIE_IN_WATCHLIST_START: 'SET_IS_MOVIE_IN_WATCHLIST_START',
  SET_IS_MOVIE_IN_WATCHLIST_SUCCESS: 'SET_IS_MOVIE_IN_WATCHLIST_SUCCESS',
  SET_IS_MOVIE_IN_WATCHLIST_FAILURE: 'SET_IS_MOVIE_IN_WATCHLIST_FAILURE',
  RATE_MOVIE: 'RATE_MOVIE',
  RATE_MOVIE_START: 'RATE_MOVIE_START',
  RATE_MOVIE_SUCCESS: 'RATE_MOVIE_SUCCESS',
  RATE_MOVIE_FAILURE: 'RATE_MOVIE_FAILURE',
};
