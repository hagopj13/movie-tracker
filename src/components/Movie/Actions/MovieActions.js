// @flow
import React, { useState, useRef } from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { makeStyles } from '@material-ui/core/styles';
import FavoriteIcon from '@material-ui/icons/Favorite';
import BookmarkIcon from '@material-ui/icons/Bookmark';
import StarRateIcon from '@material-ui/icons/StarRate';
import { red, yellow, green } from '@material-ui/core/colors';

import RatingPopover from 'components/Rating/Popover/RatingPopover';
import movieActions from 'store/movie/movie.actions';
import movieSelectors from 'store/movie/movie.selectors';
import authSelectors from 'store/auth/auth.selectors';
import dialogActions from 'store/ui/dialog/dialog.actions';
import dialogTypes from 'config/dialogTypes';
import type { Movie, MovieUserState } from 'types';

import MovieActionButton from './Button/MovieActionButton';

type Props = {
  movie: Movie,
  userState: MovieUserState,
  isAuth: boolean,
  onSetIsMovieFavorite: (id: string, isFavorite: boolean) => void,
  onSetIsMovieInWatchlist: (id: string, isInWatchlist: boolean) => void,
  onRateMovie: (id: string, rating: number | null) => void,
  onShowLoginDialog: () => void,
};

const useStyles = makeStyles((theme) => ({
  buttonsContainer: {
    '& > *': {
      marginRight: theme.spacing(1.5),
    },
  },
  favoriteIcon: {
    fontSize: 24,
    color: (props) => (props.userState.isFavorite ? red[600] : theme.palette.common.white),
  },
  watchlistIcon: {
    fontSize: 24,
    color: (props) => (props.userState.isInWatchlist ? green[500] : theme.palette.common.white),
  },
  ratingIcon: {
    fontSize: 28,
    color: (props) => (props.userState.rating ? yellow[600] : theme.palette.common.white),
  },
  tooltip: {
    fontSize: 14,
  },
}));

export const MovieActions = (props: Props) => {
  const {
    movie,
    userState,
    isAuth,
    onSetIsMovieFavorite,
    onSetIsMovieInWatchlist,
    onRateMovie,
    onShowLoginDialog,
  } = props;

  const [isRatingPopoverOpen, setIsRatingPopoverOpen] = useState(false);
  const ratingButtonRef = useRef(null);
  const classes = useStyles(props);

  const handleFavoriteButtonClick = () => {
    if (isAuth) {
      onSetIsMovieFavorite(movie.id, !userState.isFavorite);
    } else {
      onShowLoginDialog();
    }
  };

  const handleWatchlistButtonClick = () => {
    if (isAuth) {
      onSetIsMovieInWatchlist(movie.id, !userState.isInWatchlist);
    } else {
      onShowLoginDialog();
    }
  };

  const handleRatingButtonClick = () => {
    if (isAuth) {
      setIsRatingPopoverOpen(true);
    } else {
      onShowLoginDialog();
    }
  };

  const handleRatingPopoverClose = () => setIsRatingPopoverOpen(false);

  const handleRatingChange = (newValue: number | null) => {
    onRateMovie(movie.id, newValue);
  };

  const getFavoriteTooltipTitle = () => {
    if (!isAuth) {
      return 'Login to add movie to your favorite list';
    }
    if (userState.isFavorite) {
      return 'Remove movie from your favorite list';
    }
    return 'Add movie to your favorite list';
  };

  const getWatchlistTooltipTitle = () => {
    if (!isAuth) {
      return 'Login to add movie to your watchlist';
    }
    if (userState.isInWatchlist) {
      return 'Remove movie from your watchlist';
    }
    return 'Add movie to your watchlist';
  };

  const getRateTooltipTitle = () => {
    if (!isAuth) {
      return 'Login to rate movie';
    }
    if (userState.rating) {
      return `Rated ${userState.rating}`;
    }
    return 'Rate movie';
  };

  return (
    <>
      <div className={classes.buttonsContainer}>
        <MovieActionButton onClick={handleFavoriteButtonClick} title={getFavoriteTooltipTitle()}>
          <FavoriteIcon className={classes.favoriteIcon} />
        </MovieActionButton>
        <MovieActionButton onClick={handleWatchlistButtonClick} title={getWatchlistTooltipTitle()}>
          <BookmarkIcon className={classes.watchlistIcon} />
        </MovieActionButton>
        <MovieActionButton
          ref={ratingButtonRef}
          onClick={handleRatingButtonClick}
          title={getRateTooltipTitle()}
        >
          <StarRateIcon className={classes.ratingIcon} />
        </MovieActionButton>
      </div>
      <RatingPopover
        isOpen={isRatingPopoverOpen}
        anchorEl={ratingButtonRef.current}
        initialValue={userState.rating}
        onClose={handleRatingPopoverClose}
        onChange={handleRatingChange}
      />
    </>
  );
};

const mapStateToProps = createStructuredSelector({
  userState: movieSelectors.selectMovieUserState,
  isAuth: authSelectors.selectIsAuth,
});

const mapDispatchToProps = {
  onSetIsMovieFavorite: movieActions.setIsMovieFavorite,
  onSetIsMovieInWatchlist: movieActions.setIsMovieInWatchlist,
  onRateMovie: movieActions.rateMovie,
  onShowLoginDialog: () => dialogActions.showDialog(dialogTypes.LOGIN),
};

export default connect(mapStateToProps, mapDispatchToProps)(MovieActions);
