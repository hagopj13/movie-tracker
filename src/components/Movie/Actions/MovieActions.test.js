import React from 'react';
import { createShallow } from '@material-ui/core/test-utils';

import { details as movie } from 'store/fixtures/movie';
import RatingPopover from 'components/Rating/Popover/RatingPopover';

import { MovieActions } from './MovieActions';
import MovieActionButton from './Button/MovieActionButton';

describe('MovieActions component', () => {
  let shallow;
  let userState;
  let wrapper;

  const mockSetIsMovieFavorite = jest.fn();
  const mockSetIsMovieInWatchlist = jest.fn();
  const mockRateMovie = jest.fn();
  const mockShowLoginDialog = jest.fn();

  beforeAll(() => {
    shallow = createShallow();
  });

  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe('isAuth is true', () => {
    beforeEach(() => {
      userState = {
        isFavorite: false,
        isInWatchlist: false,
        rating: null,
      };
      const mockProps = {
        movie,
        userState,
        isAuth: true,
        onSetIsMovieFavorite: mockSetIsMovieFavorite,
        onSetIsMovieInWatchlist: mockSetIsMovieInWatchlist,
        onRateMovie: mockRateMovie,
        onShowLoginDialog: mockShowLoginDialog,
      };
      wrapper = shallow(<MovieActions {...mockProps} />);
    });

    it('should render the MovieActions component correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should call onSetIsMovieFavorite when the favorite button is clicked', () => {
      wrapper.find(MovieActionButton).at(0).simulate('click');
      expect(mockSetIsMovieFavorite).toHaveBeenCalledTimes(1);
      expect(mockSetIsMovieFavorite).toHaveBeenLastCalledWith(movie.id, !userState.isFavorite);
    });

    it('should call onSetIsMovieInWatchlist when the watchlist button is clicked', () => {
      wrapper.find(MovieActionButton).at(1).simulate('click');
      expect(mockSetIsMovieInWatchlist).toHaveBeenCalledTimes(1);
      expect(mockSetIsMovieInWatchlist).toHaveBeenLastCalledWith(
        movie.id,
        !userState.isInWatchlist,
      );
    });

    it('should initially have the RatingPopover closed', () => {
      expect(wrapper.find(RatingPopover).prop('isOpen')).toBe(false);
    });

    it('should open the RatingPopover when the rating button is clicked', () => {
      wrapper.find(MovieActionButton).at(2).simulate('click');
      expect(wrapper.find(RatingPopover).prop('isOpen')).toBe(true);
    });

    it('should close the RatingPopover when its onClose prop is called', () => {
      wrapper.find(MovieActionButton).at(2).simulate('click');
      wrapper.find(RatingPopover).prop('onClose')();
      expect(wrapper.find(RatingPopover).prop('isOpen')).toBe(false);
    });

    it('should call onRateMovie when RatingPopover onChange is clicked', () => {
      const newRating = 5;
      wrapper.find(RatingPopover).prop('onChange')(newRating);
      expect(mockRateMovie).toHaveBeenCalledTimes(1);
      expect(mockRateMovie).toHaveBeenLastCalledWith(movie.id, newRating);
    });
  });

  describe('isAuth is false', () => {
    beforeEach(() => {
      userState = {
        isFavorite: false,
        isInWatchlist: false,
        rating: null,
      };
      const mockProps = {
        movie,
        userState,
        isAuth: false,
        onSetIsMovieFavorite: mockSetIsMovieFavorite,
        onSetIsMovieInWatchlist: mockSetIsMovieInWatchlist,
        onRateMovie: mockRateMovie,
        onShowLoginDialog: mockShowLoginDialog,
      };
      wrapper = shallow(<MovieActions {...mockProps} />);
    });

    it('should render the MovieActions component correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should open the login dialog when favorite button is clicked', () => {
      wrapper.find(MovieActionButton).at(0).simulate('click');
      expect(mockShowLoginDialog).toHaveBeenCalledTimes(1);
    });

    it('should open the login dialog when the watchlist button is clicked', () => {
      wrapper.find(MovieActionButton).at(1).simulate('click');
      expect(mockShowLoginDialog).toHaveBeenCalledTimes(1);
    });

    it('should open the login dialog when the rating button is clicked', () => {
      wrapper.find(MovieActionButton).at(2).simulate('click');
      expect(mockShowLoginDialog).toHaveBeenCalledTimes(1);
    });
  });
});
