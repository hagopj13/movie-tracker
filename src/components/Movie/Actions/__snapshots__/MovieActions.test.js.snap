// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieActions component isAuth is false should render the MovieActions component correctly 1`] = `
<Fragment>
  <div
    className="makeStyles-buttonsContainer-1"
  >
    <ForwardRef(MovieActionButton)
      onClick={[Function]}
      title="Login to add movie to your favorite list"
    >
      <FavoriteIcon
        className="makeStyles-favoriteIcon-2 makeStyles-favoriteIcon-27"
      />
    </ForwardRef(MovieActionButton)>
    <ForwardRef(MovieActionButton)
      onClick={[Function]}
      title="Login to add movie to your watchlist"
    >
      <BookmarkIcon
        className="makeStyles-watchlistIcon-3 makeStyles-watchlistIcon-28"
      />
    </ForwardRef(MovieActionButton)>
    <ForwardRef(MovieActionButton)
      onClick={[Function]}
      title="Login to rate movie"
    >
      <StarRateIcon
        className="makeStyles-ratingIcon-4 makeStyles-ratingIcon-29"
      />
    </ForwardRef(MovieActionButton)>
  </div>
  <RatingPopover
    anchorEl={null}
    initialValue={null}
    isOpen={false}
    onChange={[Function]}
    onClose={[Function]}
  />
</Fragment>
`;

exports[`MovieActions component isAuth is true should render the MovieActions component correctly 1`] = `
<Fragment>
  <div
    className="makeStyles-buttonsContainer-1"
  >
    <ForwardRef(MovieActionButton)
      onClick={[Function]}
      title="Add movie to your favorite list"
    >
      <FavoriteIcon
        className="makeStyles-favoriteIcon-2 makeStyles-favoriteIcon-6"
      />
    </ForwardRef(MovieActionButton)>
    <ForwardRef(MovieActionButton)
      onClick={[Function]}
      title="Add movie to your watchlist"
    >
      <BookmarkIcon
        className="makeStyles-watchlistIcon-3 makeStyles-watchlistIcon-7"
      />
    </ForwardRef(MovieActionButton)>
    <ForwardRef(MovieActionButton)
      onClick={[Function]}
      title="Rate movie"
    >
      <StarRateIcon
        className="makeStyles-ratingIcon-4 makeStyles-ratingIcon-8"
      />
    </ForwardRef(MovieActionButton)>
  </div>
  <RatingPopover
    anchorEl={null}
    initialValue={null}
    isOpen={false}
    onChange={[Function]}
    onClose={[Function]}
  />
</Fragment>
`;
